app/gui/suggestion-database/src/documentation_ir.rs

```diff
@@ -75,51 +93,151 @@ impl EntryDocumentation {
+                        }
+                        .into(),
+                    });
+                    let parent_module = LinkedDocPage {
+                        name: module_docs.name.clone_ref(),
+                        page: Documentation::Module(module_docs.clone_ref()).into(),
+                    };
+                    methods.chain(constructors).chain(iter::once(parent_module)).collect()
+                }
+                Documentation::Constructor { type_docs, module_docs, .. } => vec![LinkedDocPage {
+                    name: type_docs.name.clone_ref(),
+                    page: Documentation::Type {
+                        docs:        type_docs.clone_ref(),
+                        module_docs: module_docs.clone_ref(),
+                    }
+                    .into(),
+                }],
+                Documentation::Method { type_docs, module_docs, .. } => vec![LinkedDocPage {
+                    name: type_docs.name.clone_ref(),
+                    page: Documentation::Type {
+                        docs:        type_docs.clone_ref(),
+                        module_docs: module_docs.clone_ref(),
+                    }
+                    .into(),
+                }],
+                Documentation::ModuleMethod { module_docs, .. } => vec![LinkedDocPage {
+                    name: module_docs.name.clone_ref(),
+                    page: Documentation::Module(module_docs.clone_ref()).into(),
+                }],
+                Documentation::Function(_) => default(),
+                Documentation::Local(_) => default(),
+                Documentation::Builtin(_) => default(),
+            },
+            EntryDocumentation::Placeholder(_) => default(),
         }
     }
 
-    /// Create documentation for a hard-coded builtin entry.
-    pub fn builtin(sections: impl IntoIterator<Item = &DocSection>) -> Self {
-        let sections = Rc::new(BuiltinDocumentation::from_doc_sections(sections.into_iter()));
-        Self::Docs(Documentation::Builtin(sections))
+    fn parent_module(
+        db: &SuggestionDatabase,
+        entry: &Entry,
+        entry_id: entry::Id,
+    ) -> Result<ModuleDocumentation, NoParentModule> {
+        let defined_in = &entry.defined_in;
+        let parent_module = db.lookup_by_qualified_name(defined_in);
+        match parent_module {
+            Some((id, parent)) => match parent.kind {
+                Kind::Module => Ok(ModuleDocumentation::new(id, &parent, db)
+                    .map_err(|_| NoParentModule(entry_id))?),
+                _ => {
+                    error!(
+                        "Unexpected parent entry ({}) kind, expected Module.",
+                        entry.qualified_name()
+                    );
+                    Err(NoParentModule(entry_id))
+                }
+            },
+            None => {
+                error!("Parent module for entry {} not found.", entry.qualified_name());
+                Err(NoParentModule(entry_id))
```

Comment:
Here we both print error and return `Err` - but the caller should handler `Err` possibly printing the error again. I think we don't need to log errors in functions returning `Err`.

Reply:
Yep, you're right. Removed the logs and improved the error message.